### PR TITLE
PYIC-1450 Validate JTI of access token private key JWT request

### PR DESCRIPTION
**Pull request by blakeyp**
### What changed

Validate the JWT id (JTI) of the access token private key JWT request
- check that it is present in the JWT claims
- check that it is unique ie. has not already been used before, by storing previously used JWT ids

### Why did it change

OpenID Connect Core 1.0 Section 9 ([Final: OpenID Connect Core 1.0 incorporating errata set 1](https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication)) states that the private_key_jwt is required to contain a jti claim, which is a unique identifier for the JWT. It should be used to prevent re-use of the tokens, which MUST only be used once

### Issue tracking
- [PYIC-1450](https://govukverify.atlassian.net/browse/PYIC-1450)

### Environment variables or secrets
- [X] No environment variables or secrets were added or changed

### CODE REVIEW

File: lambdas/accesstoken/src/main/java/uk/gov/di/ipv/cri/passport/accesstoken/AccessTokenHandler.java [hunk 1]
```diff
@@ -63,7 +63,7 @@ public APIGatewayProxyResponseEvent handleRequest(
             APIGatewayProxyRequestEvent input, Context context) {
         LogHelper.attachComponentIdToLogs();
         try {
-            tokenRequestValidator.authenticateClient(input.getBody());
+            String jwtId = tokenRequestValidator.authenticateClient(input.getBody());
```
Comment (Wynndow) [inline lambdas/accesstoken/src/main/java/uk/gov/di/ipv/cri/passport/accesstoken/AccessTokenHandler.java:66]: Could this var be called something else, like clientAuthJwtId, just to make it a little clearer where this ID comes from

**Review by Wynndow — COMMENTED** [review]
This looks great!
Now that I've seen how it might work with storing the jti on the access token, I actually think it might be better to store it in its own table instead. The way it is at the moment, if anything goes wrong before we store the access token in the AccessTokenHandler and we return an error, the jti will be available to be re-used. I feel like we need to store it/check it when we do the authentication. And we can't store it on the access token and that point as we haven't created it yet. Let's have a chat about it today.

File: lambdas/accesstoken/src/main/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidator.java [hunk 2]
```diff
@@ -66,6 +78,24 @@ private void validateMaxAllowedAuthClientTtl(JWTAuthenticationClaimsSet claimsSe
...
+        JWTID jwtId = claimsSet.getJWTID();
+        if (jwtId == null || StringUtils.isBlank(jwtId.getValue())) {
+            LOGGER.error("The client auth JWT id (jti) is missing");
+            throw new InvalidClientException("The client auth JWT id (jti) is missing");
+        }
+        ClientAuthJwtIdItem clientAuthJwtIdItem =
+                clientAuthJwtIdService.getClientAuthJwtIdItem(jwtId.getValue());
+        if (clientAuthJwtIdItem != null) {
+            LoggingUtils.appendKey("jti", jwtId.getValue());
+            LoggingUtils.appendKey("usedAtDateTime", clientAuthJwtIdItem.getUsedAtDateTime());
```
Comment (Wynndow) [inline lambdas/accesstoken/src/main/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidator.java:91]: Might be nice to make these keys constants

File: lambdas/accesstoken/src/test/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidatorTest.java [hunk 3]
```diff
@@ -45,14 +48,17 @@ class TokenRequestValidatorTest {
 
     private TokenRequestValidator validator;
     @Mock private ConfigurationService mockConfigurationService;
+    @Mock private ClientAuthJwtIdService mockClientAuthJwtIdService;
```
Comment (Wynndow) [inline lambdas/accesstoken/src/test/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidatorTest.java:51]:
```suggestion
    @Mock private ConfigurationService mockConfigurationService;
    @Mock private ClientAuthJwtIdService mockClientAuthJwtIdService;
    @InjectMocks private TokenRequestValidator validator;
```

This way you can drop the instantiation in the setup method

File: lib/src/test/java/uk/gov/di/ipv/cri/passport/library/service/ClientAuthJwtIdServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+public class ClientAuthJwtIdServiceTest {
+    @Mock private DataStore<ClientAuthJwtIdItem> mockDataStore;
+
+    @Mock private ConfigurationService mockConfigurationService;
+
+    private ClientAuthJwtIdService clientAuthJwtIdService;
```
Comment (Wynndow) [inline lib/src/test/java/uk/gov/di/ipv/cri/passport/library/service/ClientAuthJwtIdServiceTest.java:25]:
```suggestion
    @InjectMocks private ClientAuthJwtIdService clientAuthJwtIdService;
```
Then you can delete the setUp method

File: deploy/template.yaml [hunk 5]
```diff
@@ -522,6 +523,8 @@ Resources:
       AttributeDefinitions:
         - AttributeName: "accessToken"
           AttributeType: "S"
+        - AttributeName: "jwtId"
+          AttributeType: "S"
```
[the remark this answers is not in the document]
Comment (blakeyp) [reply deploy/template.yaml]: Good spot, removed

File: lambdas/accesstoken/src/main/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidator.java (hunk 2 above)
Comment (blakeyp) [reply lambdas/accesstoken/src/main/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidator.java]: Added to `LogField` enum

File: lib/src/test/java/uk/gov/di/ipv/cri/passport/library/service/ClientAuthJwtIdServiceTest.java (hunk 4 above)
Comment (blakeyp) [reply lib/src/test/java/uk/gov/di/ipv/cri/passport/library/service/ClientAuthJwtIdServiceTest.java]: Didn't know you could do this, nice! Done.

File: lambdas/accesstoken/src/test/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidatorTest.java (hunk 3 above)
Comment (blakeyp) [reply lambdas/accesstoken/src/test/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidatorTest.java:51]: Having some trouble with this one as it also needs the `verifier` injecting, but seems we are exercising the real implementation of `verifier` in the tests so I don't think we want it mocked?

Comment (Wynndow) [reply lambdas/accesstoken/src/test/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidatorTest.java:51]: Oh interesting, I missed the fact that we're creating the verifier in the constructor.

**Review by Wynndow — APPROVED** [review]

**Review by Wynndow — APPROVED** [review]

**Pull request merged**